### ISSUE TITLE
flutter pub get : Authentication error (403)

### ISSUE THREAD

**Issue by vermefr** [dart-lang/pub#3662]
Hi, 

I get an error when doing a flutter pub get, under Linux (Ubuntu 22.04.1 LTS).
I've attached my pub.log.txt file.

Many thanks in advance :)

Console Output:
```
$ flutter pub get
Running "flutter pub get" in flutter_tools...                    2 751ms
Resolving dependencies in ../../flutter/packages/flutter_tools...
Authentication error (403)
package:pub/src/authentication/client.dart 95:5    _AuthenticatedClient._throwAuthException
package:pub/src/authentication/client.dart 60:9    _AuthenticatedClient.send
===== asynchronous gap ===========================
package:pub/src/authentication/client.dart 133:12  withAuthenticatedClient
===== asynchronous gap ===========================
package:pub/src/source/hosted.dart 1056:28         HostedSource._download.<fn>.<fn>
===== asynchronous gap ===========================
package:pub/src/utils.dart 703:14                  retry
===== asynchronous gap ===========================
package:pub/src/source/hosted.dart 1052:7          HostedSource._download.<fn>
===== asynchronous gap ===========================
package:pub/src/io.dart 886:12                     withTempDir
===== asynchronous gap ===========================
package:pub/src/source/hosted.dart 1008:12         HostedSource._download
===== asynchronous gap ===========================
package:pub/src/source/hosted.dart 758:21          HostedSource.downloadToSystemCache
===== asynchronous gap ===========================
package:pub/src/system_cache.dart 230:12           SystemCache.downloadPackage
===== asynchronous gap ===========================
package:pub/src/solver/result.dart 70:20           SolveResult.downloadCachedPackages.<fn>.<fn>
===== asynchronous gap ===========================
package:pub/src/solver/result.dart 68:18           SolveResult.downloadCachedPackages.<fn>
===== asynchronous gap ===========================
package:pub/src/solver/result.dart 65:32           SolveResult.downloadCachedPackages
===== asynchronous gap ===========================
package:pub/src/entrypoint.dart 352:25             Entrypoint.acquireDependencies
===== asynchronous gap ===========================
age:pub/src/command/get.dart 53:5              GetCommand.runProtected
This is an unexpected error. The full log and other details are collected in:

    /home/verme/.pub-cache/log/pub_log.txt

Consider creating an issue on https://github.com/dart-lang/pub/issues/new
and attaching the relevant parts of that log file.
pub get failed
command: "/home/verme/flutter/bin/cache/dart-sdk/bin/dart __deprecated_pub --color --directory
../../flutter/packages/flutter_tools get --example"
pub env: {
  "FLUTTER_ROOT": "/home/verme/flutter",
  "PUB_ENVIRONMENT": "flutter_cli:get",
  "PUB_CACHE": "/home/verme/.pub-cache",
[pub_log.txt](https://github.com/dart-lang/pub/files/10069099/pub_log.txt)

}
exit code: 1
last line of pub output: "and attaching the relevant parts of that log file."

```

**Comment by vermefr** [comment]
My log file:

[pub_log.txt](https://github.com/dart-lang/pub/files/10069126/pub_log.txt)

**Comment by vermefr** [comment]
Additional comments: I can access URLs from a navigator, and I am not using a proxy nor a VPN.

**Comment by vermefr** [comment]
I am not using a pub mirror.

My request is done from Switzerland.

I'm doing a "flutter pub get".

Flutter & Dart Versions : 
Flutter 3.6.0-0.1.pre • channel beta • https://github.com/flutter/flutter.git
Framework • revision 75927305ff (il y a 6 jours) • 2022-11-17 05:52:50 +0530
Engine • revision b754eabc98
Tools • Dart 2.19.0 (build 2.19.0-374.1.beta) • DevTools 2.19.0

My pubspec.yaml :
[pubspec.zip](https://github.com/dart-lang/pub/files/10074344/pubspec.zip)

**Comment by sigurdm** [comment]
This looks weird

Does this happen consistently?

I can see in the log-file that:
```
IO  : HTTP response 200 OK for GET https://pub.dev/api/packages/meta
```
but
```
IO  : HTTP response 403 Forbidden for GET https://pub.dartlang.org/packages/file_testing/versions/3.0.0.tar.gz
```
So it seems the package-listings are going through fine, but not the package downloads.

@isoos , @jonasfj any ideas?

**Comment by sigurdm** [comment]
This could make sense from a country with a blocking policy, but from Switzerland it should work, and also it is weird that the curl command works.

**Comment by vermefr** [comment]
Yes VERY weird indeed !

"flutter pub get" has been working for a long time; until yesterday. 
Since, I can't execute this command successfully anymore.

My Last findings:
1. dart pub cache clean => /home/verme/.pub-cache    folder gets empty.
2. dart pub get => /home/verme/.pub-cache    is filled with attached contents:

[pub-cache.zip](https://github.com/dart-lang/pub/files/10074601/pub-cache.zip)

**Comment by isoos** [comment]
`curl -v https://pub.dartlang.org/packages/file_testing/versions/3.0.0.tar.gz` returns with  `location: https://storage.googleapis.com/dartlang-pub-public-packages/packages/file_testing-3.0.0.tar.gz` and then it just downloads fine for me. Is there any authentication token set that may be sent to `storage.googleapis.com`?

**Comment by sigurdm** [comment]
`dart pub token list`

**Comment by jonasfj** [comment]
@vermefr 

okay, so that sounds like it's not an issue in `pub` or `pub.dev`, but rather GCS.

Also it's not deterministic, it's intermittent since you do have successfully downloaded some packages, right? You pub-cache is not empty.

**Comment by vermefr** [comment]
My IP geolocation:

IP ADDRESS: 195.15.222.163
 COUNTRY: Switzerland 
 REGION: Geneva
 CITY: Onex
 ISP: [Infomaniak](http://ipinfo.io/AS29222)
 ORGANIZATION: Infomaniak Network SA ([infomaniak.com](https://host.io/infomaniak.com))

**Comment by vermefr** [comment]
> @vermefr
> 
> okay, so that sounds like it's not an issue in `pub` or `pub.dev`, but rather GCS.
> 
> Also it's not deterministic, it's intermittent since you do have successfully downloaded some packages, right? You pub-cache is not empty.

Yes, some packages are downloaded successfully; but not all. And this has been a permanent situation since yesterday.

**Comment by jonasfj** [comment]
Ah, you had posted the XML, just not using code-fences, so we couldn't read it -- had to edit the message.

quick note: Please use [code fences](https://help.github.com/en/github/writing-on-github/creating-and-highlighting-code-blocks) when posting logs, for large log files please use a [gist](https://gist.github.com/).

**Comment by vermefr** [comment]
I've just disabled ipv6, and yeah !! it works now !

MANY thanks to all of you 🙂👍👍

**Comment by vermefr** [comment]
Is there still something to fix in GCS about ipv6 geolocation ?

**Comment by jonasfj** [comment]
@vermefr I don't know, how does your ipv6 show up?
If you have time feel free to drop me an email with your ipv6 address and config information, then I'll see if I can find anyone to ask. Hit me at jonasfj@google.com

But I'm not exactly an ipv6 expert, and I guess it's hard to tell if it's a temporary ISP issue? or something else.

**Comment by vermefr** [comment]
I've answered you by email :)

**Comment by anas2027** [comment]
do you find the soultion to your problem plz help me its happend me just when i import sharedPrefence

**Comment by vermefr** [comment]
@anas2027: My problem came from ipv6 that failed to work correctly at my internet provider side.
I solved this by deactivating ipv6 on my Linux machine.

Hope this will help you :)

**Comment by sigurdm** [comment]
@jonasfj here is more evidence of your ipv6 configuration theory.
